core/src/HIP/Kokkos_HIP_Parallel_Team.hpp

```diff
@@ -755,10 +757,34 @@ class ParallelReduce<CombinedFunctorReducerType,
                                  !m_result_ptr_host_accessible ||
                                  !std::is_same<ReducerType, InvalidType>::value;
     if (!is_empty_range || need_device_set) {
-      const int block_count =
-          UseShflReduction
-              ? std::min(m_league_size, size_type(1024 * HIPTraits::WarpSize))
-              : std::min(static_cast<int>(m_league_size), m_team_size);
+      int block_max = 0;
+      constexpr auto light_weight =
+          Kokkos::Experimental::WorkItemProperty::HintLightWeight;
+      constexpr typename Policy::work_item_property property;
+      if ((property & light_weight) == light_weight) {
+        block_max = 2097152;
+      } else {
+        block_max = 65536;
+      }
+      constexpr int preferred_block_min = 1024;
+      int block_count_tmp               = m_league_size;
+      if (block_count_tmp < preferred_block_min) {
+        // keep blocks as is, already low parallelism
+      } else if (block_count_tmp >= block_max) {
+        block_count_tmp = block_max;
+
+      } else {
+        int nwork            = m_league_size * m_team_size;
+        int items_per_thread = (nwork + block_count_tmp * m_team_size - 1) /
+                               (block_count_tmp * m_team_size);
+        if (items_per_thread < 4) {
+          int ratio = std::min(
+              (block_count_tmp + preferred_block_min - 1) / preferred_block_min,
+              (4 + items_per_thread - 1) / items_per_thread);
+          block_count_tmp /= ratio;
+        }
+      }
```

Comment:
Wouldn't it make sense to apply the first case unconditionally, i.e,
```suggestion
      } else {
        int nwork            = m_league_size * m_team_size;
        int items_per_thread = (nwork + block_count_tmp * m_team_size - 1) /
                               (block_count_tmp * m_team_size);
        if (items_per_thread < 4) {
          int ratio = std::min(
              (block_count_tmp + preferred_block_min - 1) / preferred_block_min,
              (4 + items_per_thread - 1) / items_per_thread);
          block_count_tmp /= ratio;
        }
        if (block_count_tmp >= block_max)
          block_count_tmp = block_max;
      }
```

Reply:
What's the point of doing that? It's harder to read.

Reply:
The current logic limits the number of blocks to some maximum value if it exceeds it. If the number of blocks is already less than that, it potentially chooses an even smaller value. 
It seems to me that it makes sense to first potentially increase the number of items per thread and only if the resulting block count still exceeds `block_max` set it to `block_max`.

Reply:
I am not convinced that you would get better results. It's just an heuristic. We can try but we need to rerun the benchmark.

Reply:
Of course, benchmarks would need to be rerun. It just sticks out that the current version looks discontinuous around `block_max` whenever the number of items per thread is less than 4.